engine/cluster/src/main/java/org/opennms/alec/engine/cluster/AbstractClusterEngine.java

```diff
@@ -548,11 +548,23 @@ private String getDiagnosticTextForSituation(Situation situation) {
             getOptionalVertexIdForAlarm(alarm).ifPresent(vertexIds::add);
         }
 
+        boolean skippedDistanceCalc = false;
         if (vertexIds.size() < NUM_VERTEX_THRESHOLD_FOR_HOP_DIAG) {
             maxSpatialDistance = 0d;
             for (Long vertexIdA : vertexIds) {
                 for (Long vertexIdB : vertexIds) {
                     if (!vertexIdA.equals(vertexIdB)) {
+                        // Make sure both vertices are still on the filtered graph and if not, skip this pair
+                        CEVertex a = graphManager.getVertexWithId(vertexIdA);
```

Comment:
This could be optimized by performing the vertex filtering before entering the nested `for` loops - this way we only have to check each vertex once.

Reply:
Right you are, fixed.